contrib: Use asmap for ASN lookup in makeseeds

laanwj opened the pull request:
Add an argument `-a` to provide a asmap file to do the IP to ASN lookups.

This speeds up the script greatly, and makes the output deterministic. Also removes the dependency on `dns.lookup`.

I've annotated the output with ASxxxx comments to provide a way to verify the functionality.

For now I've added instructions in README.md to download and use the `demo.map` from the asmap repository. When we have some other mechanism for distributing asmap files we could switch to that.

This continues #24824. I've removed the fallbacks and extra complexity, as everyone will be using the same instructions anyway.

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>
Co-authored-by: russeree <reese.russell@ymail.com>

jonatack committed:
contrib: add seeds progress indicator and remove asmap one in makeseeds script

laanwj committed:
contrib: Use asmap for ASN lookup in makeseeds

Add an argument `-a` to provide a asmap file to do the IP to ASN
lookups.

This speeds up the script greatly, and makes the output deterministic.
Also removes the dependency on `dns.lookup`.

I've annotated the output with ASxxxx comments to provide a way to
verify the functionality.

For now I've added instructions in README.md to download and use the
`demo.map` from the asmap repository. When we have some other mechanism
for distributing asmap files we could switch to that.

This continues #24824. I've removed all the fallbacks and extra
complexity, as everyone will be using the same instructions anyway.

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>
Co-authored-by: James O'Beirne <james.obeirne@pm.me>
Co-authored-by: russeree <reese.russell@ymail.com>

jonatack commented:
Concept ACK. 

(Python linter says to remove unused headers.)

laanwj commented on contrib/seeds/makeseeds.py:
```diff
@@ -179,13 +155,14 @@ def filterbyasn(ips: List[Dict], max_per_asn: Dict, max_per_net: int) -> List[Di
             # ips from this ASN on this network
             continue
-        asn_count[asn] += 1
+        asn_count[ip['net'], asn] += 1
```
btw: This fixes a bug that was exposed by the changes in #24818, where it was possible for a limit to be exceeded because `max_per_asn` differs per network but `asn_count` was not counted per network (in combination with `==`). Making the counting per ipv4/ipv6 avoids this, and makes sense to me conceptually. So after this, a single ASN can have *both* 2 IPv4 seeds and 6 IPv6 seeds.

jonatack replied:
With this change I only see 12 entries in nodes_main.txt though (2 IPv4, 10 IPv6, all AS30765).

laanwj replied:
Which asmap are you using? It seems more likely to me that your asmap is wrong, or something is wrong with parsing, causing all IPs to be assigned to one AS, than that this change is at fault.

laanwj replied:
FWIW this is my map file:
```
sha256sum demo.map
94b6142b574d468e73c71ed9fb411683d295c37b756f1f735013f736ee647330  demo.map
```
Output:
```
Loading asmap database demo.map... Done.
  IPv4   IPv6  Onion Pass
470789  73296      0 Initial
470789  73296      0 Skip entries with invalid address
470789  73296      0 After removing duplicates
  6318   1747      0 Enforce minimal number of blocks
  5456   1504      0 Require service bit 1
  3866    887      0 Require minimum uptime
  3789    864      0 Require a known and recent user agent
  3760    856      0 Filter out hosts with multiple bitcoin ports
   512    301      0 Look up ASNs and limit results per ASN and per net
```

russeree replied:
I currently am having this same bug using the asmap from the readme.md included.
`curl https://github.com/sipa/asmap/raw/master/demo.map > demo.map`

```
sha256sum demo.map
72034b228f2b91d8c0864d27d5f8cbbad4b34f1b0948fd2bc9f1acf66aaf168c  demo.map
```
```
Loading asmap database demo.map... Done.
  IPv4   IPv6  Onion Pass
470888  73307      0 Initial
470888  73307      0 Skip entries with invalid address
470888  73307      0 After removing duplicates
470887  73307      0 Skip entries from suspicious hosts
  6502   1787      0 Enforce minimal number of blocks
  5610   1523      0 Require service bit 1
  3862    896      0 Require minimum uptime
  3777    869      0 Require a known and recent user agent
  3753    861      0 Filter out hosts with multiple bitcoin ports
     2     10      0 Look up ASNs and limit results per ASN and per net

real    0m11.036s
user    0m10.079s
sys     0m0.907s
```

The time to parse is insanely fast but the included file is only 2.5kb which to me seems incredibly small as an uncompressed asmap is nearly 27MB.

sipa replied:
@russeree If you use curl, it'll just store "You are being redirected to ..." in the file.

That demo.dap file is also very old, and wasn't even intended back then to be complete. Try https://bitcoin.sipa.be/asmap-filled.dat.

russeree replied:
```
cat demo.map
<html><body>You are being <a href="https://raw.githubusercontent.com/sipa/asmap/master/demo.map">redirected</a>.</body></html>
```

well then ... after using the correct file things looked much better.

`Loading asmap database demo.map... Done.
  IPv4   IPv6  Onion Pass
470888  73307      0 Initial
470888  73307      0 Skip entries with invalid address
470888  73307      0 After removing duplicates
470887  73307      0 Skip entries from suspicious hosts
  6502   1787      0 Enforce minimal number of blocks
  5610   1523      0 Require service bit 1
  3862    896      0 Require minimum uptime
  3777    869      0 Require a known and recent user agent
  3753    861      0 Filter out hosts with multiple bitcoin ports
   512    291      0 Look up ASNs and limit results per ASN and per net

real    0m15.295s
user    0m14.423s
sys     0m0.808s`

using a broken file presents an interesting note to me. When using the broken file the script just used "You are being redirected to ..." as the ASN for ipv4 with a limit of 2. Then does ipv6 also use "You are being redirected to ..." for the 10 items per ASN? So the max allocated ip per ASN is 12 including ipv4 and ipv6?

laanwj replied:
> So the max allocated ips per ASN is actually 12 including ipv4 and ipv6?

Yes. The maximum number of seeds per ASN is 12: 2 IPv4 and 10 IPv6.

jonatack replied:
> Which asmap are you using? It seems more likely to me that your asmap is wrong, or something is wrong with parsing, causing all IPs to be assigned to one AS, than that this change is at fault.

Good call; fixed in the latest push.

laanwj commented on contrib/seeds/asmap.py:
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2013-2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+import ipaddress
```
BTW I didn't realize before that Python had built-in functionality to parse/manipulate IP addresses (both v4 and v6). Might make sense (but not in this PR) to use this in more places and replace all the ad-hoc hacks.

laanwj commented on contrib/seeds/asmap.py:
```diff
@@ -0,0 +1,90 @@
+def DecodeBytes(byts):
+    return [(byt >> i) & 1 for byt in byts for i in range(8)]
+
+def DecodeBits(stream, bitpos, minval, bit_sizes):
```
The code from here came from [testmap](https://github.com/sipa/asmap/blob/master/testmap.py) which doesn't come with a license. It looks like @sipa and @jamesob contributed to it, so to include it we would need you to sign off that it's acceptable to include this here under the MIT license.

sipa replied:
ACK MIT license from me.

Also, I'm currently rewriting the Python asmap code from scratch actually (to have a single class that can do creation of asmap files, decoding, lookup, diffing, have tests, ...), and it's close to being done.

jamesob replied:
ACK MIT license

laanwj replied:
Thank you!

> Also, I'm currently rewriting the Python asmap code from scratch actually (to have a single class that can do creation of asmap files, decoding, lookup, diffing, have tests, ...), and it's close to being done.

OK, we can wait for that, there's no hurry here. It would be nice to have this before 0.24 branch-off.

dunxen replied:
> Also, I'm currently rewriting the Python asmap code from scratch actually (to have a single class that can do creation of asmap files, decoding, lookup, diffing, have tests, ...), and it's close to being done.

**/start slightly off-topic:**

Nice! Is the diffing for encoded mappings or just output from `asmap-rs`? Or would this even make `asmap-rs` obsolete? I'm in the process of making some improvements to UX there to show progress on RIPE data dump downloads and finding the ASN bottlenecks. I was also thinking of including a command there to compare non-encoded maps. I was going to use that to run some sort of monthly historical analysis over the past year to get an idea of how much the maps tend to change. Just don't want to duplicate effort.

Just some of the things to help get asmaps distributed with releases and enabled by default. @sipa, just let me know if you want to continue this chat off GitHub or on some other issue. :)

**/end slightly off-topic**

jonatack commented:
Tested that both `-a` and `--asmap` run, but if `makeseeds` no longer works without the added arg, can it be removed?

```
$ python3 makeseeds.py < seeds_main.txt > nodes_main.txt
usage: makeseeds.py [-h] -a ASMAP
makeseeds.py: error: the following arguments are required: -a/--asmap
```

jonatack commented on contrib/seeds/README.md:
```diff
@@ -11,18 +11,7 @@ to addrman with).
 The seeds compiled into the release are created from sipa's DNS seed data, like this:
```
Maybe update here
```suggestion
The seeds compiled into the release are created from sipa's DNS seed and AS map data, like this:
```

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -201,7 +178,18 @@ def ip_stats(ips: List[Dict]) -> str:
+
+    print(f'Loading asmap database {args.asmap}... ', end='', file=sys.stderr, flush=True)
+    asmap = ASMap(args.asmap)
+    print(f'Done.', file=sys.stderr)
```
a few nitty suggestions here
```diff
-    print(f'Loading asmap database {args.asmap}... ', end='', file=sys.stderr, flush=True)
+    print(f'Loading asmap database "{args.asmap}"... ', end='', file=sys.stderr, flush=True)
     asmap = ASMap(args.asmap)
-    print(f'Done.', file=sys.stderr)
+    print('done.\n', file=sys.stderr)
```

laanwj commented:
> Tested that both -a and --asmap run, but if makeseeds no longer works without the added arg, can it be removed?

I considered this, but I have a slight preference for explicit named instead of positional arguments that's why I kept the `-a`/`--asmap`.

jonatack commented:
> I considered this, but I have a slight preference for explicit named instead of positional arguments that's why I kept the `-a`/`--asmap`.

Ah, makes sense.

jonatack commented on contrib/seeds/makeseeds.py:
```diff
@@ -201,7 +178,18 @@ def ip_stats(ips: List[Dict]) -> str:
 
+def parse_args():
+    argparser = argparse.ArgumentParser(description=f'Generates a list of bitcoin node seed ip addresses.')
+    argparser.add_argument("-a","--asmap", help=f'The location of the asmap asn database file (required)', required=True)
```
nits
```diff
-    argparser = argparse.ArgumentParser(description=f'Generates a list of bitcoin node seed ip addresses.')
-    argparser.add_argument("-a","--asmap", help=f'The location of the asmap asn database file (required)', required=True)
+    argparser = argparse.ArgumentParser(description='Generates a list of bitcoin node seed ip addresses.')
+    argparser.add_argument("-a", "--asmap", help='the location of the asmap asn database file (required)', required=True)
```

```
$  python3 makeseeds.py -h
usage: makeseeds.py [-h] -a ASMAP

Generates a list of bitcoin node seed ip addresses.

optional arguments:
  -h, --help            show this help message and exit
  -a ASMAP, --asmap ASMAP
                        the location of the asmap asn database file (required)
```

sipa commented:
I've put more up-to-date and generic asmap files on https://bitcoin.sipa.be/asmap-unfilled.dat and https://bitcoin.sipa.be/asmap-filled.dat (the latter one has subnets with no actual ASN assigned to nearby ASNs to minimize the file size). The input data was sourced through https://github.com/rrybarczyk/asmap-rs, and compiled to asmap format by new code I'm working on at https://github.com/sipa/asmap/tree/nextgen. I hope that's in a presentable state in the next few days somewhere, but I'm put it online already to experiment with.

laanwj commented:
Re-pushed:
- Addressed @jonatack 's comments
- Changed curl URL to @sipa's new "asmap-filled.dat" asmap database (sorry for providing link to a corrupted one)
- Mention @jamesob as co-author

jonatack commented on contrib/seeds/README.md:
```diff
@@ -8,21 +8,11 @@ and remove old versions as necessary (at a minimum when GetDesirableServiceFlags
 
-The seeds compiled into the release are created from sipa's DNS seed data, like this:
+The seeds compiled into the release are created from sipa's DNS seed and AS map
+data, like this:
```
nit, maybe be explicit regarding the directory to run these commands from
```suggestion
data. Run the following commands from the `/contrib/seeds` directory:
```

laanwj replied:
Makes sense.

jonatack commented on contrib/seeds/README.md:
```diff
@@ -8,21 +8,11 @@ and remove old versions as necessary (at a minimum when GetDesirableServiceFlags
 
     curl https://bitcoin.sipa.be/seeds.txt.gz | gzip -dc > seeds_main.txt
-    python3 makeseeds.py < seeds_main.txt > nodes_main.txt
+    curl https://bitcoin.sipa.be/asmap-filled.dat > asmap-filled.dat
```
Should this filename be made consistent (one way or the other and .gitignore too) with `DEFAULT_ASMAP_FILENAME`?
```cpp
src/init.cpp:129:static const char* DEFAULT_ASMAP_FILENAME="ip_asn.map";
```

laanwj replied:
I don't know, I think that use of asmap is completely unrelated. Using the same name would make sure if we were to use the one from the user's bitcoin directory. But we want to use a specific, deterministic asmap here so everyone gets the same result.

jonatack replied:
Good point, agree.

jonatack reviewed (commented):
Lightly tested ACK 4f331939706fddd210d910ed71a9368747a48dae per `git range-diff 10a626a1 08ecc04 4f33193` 

```
Loading asmap database "asmap-filled.dat"…Done.
  IPv4   IPv6  Onion Pass                                               
470769  73264      0 Initial
470769  73264      0 Skip entries with invalid address
470769  73264      0 After removing duplicates
  6321   1728      0 Enforce minimal number of blocks
  5492   1496      0 Require service bit 1
  3861    872      0 Require minimum uptime
  3782    849      0 Require a known and recent user agent
  3757    841      0 Filter out hosts with multiple bitcoin ports
   512    281      0 Look up ASNs and limit results per ASN and per net
```

laanwj commented:
Cherry-picked a commit from @jonatack with improvements.

jonatack commented:
re-ACK 7edacdc6952a4327b01dec34d664450f055a10d0

dunxen reviewed (approved):
tACK 7edacdc

With this version of the map:

```shell
₿ sha256sum asmap-filled.dat 
f479906ce1731281616a235a85a79c9c5085c36d88689939ef7fcc5196d30874  asmap-filled.dat
```

I get:

```shell
₿ python3 makeseeds.py -a asmap-filled.dat < seeds_main.txt > nodes_main.txt
Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion Pass                                               
471267  73410      0 Initial
471267  73410      0 Skip entries with invalid address
471267  73410      0 After removing duplicates
  6927   1908      0 Enforce minimal number of blocks
  5967   1617      0 Require service bit 1
  3636    856      0 Require minimum uptime
  3244    771      0 Require a known and recent user agent
  3221    767      0 Filter out hosts with multiple bitcoin ports
   512    238      0 Look up ASNs and limit results per ASN and per net
```

laanwj commented:
I'm kind of in doubt here, with the testing and review this has got, should we go ahead and merge this and leave updating to `asmap-nextgen` for a follow-up PR?

dunxen commented:
> should we go ahead and merge this and leave updating to `asmap-nextgen` for a follow-up PR?

I think this is fair :)
Definitely adds a lot of value already.

sipa commented:
The current module there in asmap.py has an ASMap class with tests, and a from_binary and lookup method, which should be sufficient for the purposes here. I don't think much is going to change there, and it's already much better at handling error conditions than the current code, so feel free to switch to that.

I'm going to still clean things up more and have some cli tools using it, but the asmap module is pretty much done.

w0xlt reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/24864/commits/7edacdc6952a4327b01dec34d664450f055a10d0

laanwj committed:
contrib: Update makeseeds to asmap-nextgen

laanwj commented:
> I'm going to still clean things up more and have some cli tools using it, but the asmap module is pretty much done.

I've replaced `asmap.py` with the new one. I didn't find a way to look up an IP address directly so I did need to port over the `decode_ip` function. Please let me know if anything can be done more efficiently.

Result is the same, but it's a bit slower (I think the bottleneck is in the loading phase):

Before:
```
$ time python3 makeseeds.py -a asmap-filled.dat < seeds_main.txt > /dev/null 
Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion Pass                                               
470789  73296      0 Initial
470789  73296      0 Skip entries with invalid address
470789  73296      0 After removing duplicates
  6318   1747      0 Enforce minimal number of blocks
  5456   1504      0 Require service bit 1
  3866    887      0 Require minimum uptime
  3789    864      0 Require a known and recent user agent
  3760    856      0 Filter out hosts with multiple bitcoin ports
   512    295      0 Look up ASNs and limit results per ASN and per net

real    0m9.826s
user    0m9.349s
sys     0m0.468s
```

After:
```
Loading asmap database "asmap-filled.dat"…Done.
Loading and parsing DNS seeds…Done.
  IPv4   IPv6  Onion Pass                                               
470789  73296      0 Initial
470789  73296      0 Skip entries with invalid address
470789  73296      0 After removing duplicates
  6318   1747      0 Enforce minimal number of blocks
  5456   1504      0 Require service bit 1
  3866    887      0 Require minimum uptime
  3789    864      0 Require a known and recent user agent
  3760    856      0 Filter out hosts with multiple bitcoin ports
   512    295      0 Look up ASNs and limit results per ASN and per net

real    0m35.717s
user    0m35.050s
sys     0m0.628s
```

sipa commented:
@laanwj Ah yes, there are only functions for converting between net ranges and prefixes, not addresses. I'll add functions for those too. I don't think that's holding up anything here though; I can PR an update when it's done.

The slowdown is indeed somewhat expected, as it's converting the whole map to a lookup tree format rather than interpreting the binary data directly. Is that a concern? I could add interpretation based logic too.

laanwj commented:
> Is that a concern? I could add interpretation based logic too.

No, not for me at least. It's not a script to be run often and ~30s still a lot better than when it had to DNS query for every address. I also think the error checking implied by parsing instead of interpretation is useful here.

jonatack commented:
Between 17 and 20 seconds total `makeseeds.py` time for me, which seems fine. The slower parts are preceded by a "Loading..." message, so all good.

laanwj commented:
Re-pushed to fix linter errors (upstream in sipa/asmap#5).

However there's still one left which I don't really know how to get rid of:
```
contrib/seeds/asmap.py:5:1: F407 future feature annotations is not defined
```
I don't get this locally. It must have to do with Python version differences.

sipa commented:
FWIW the asmap code runs significantly faster inside of `pypy3`.

jonatack commented:
Was just wondering about this. Wolud rebasing help, to use the python instead of bash python linter?

laanwj commented:
I added a commit that disables the Python linter for `asmap.py`.
```
    test: Exclude asmap.py from Python linter checks
    
    asmap.py, a third-party dependency, is excluded from Python linter checks
    because of `from __future__ import annotations` (PEP 563 postponed annotations)
    this workaround can be removed when the Python version used is upgraded to 3.7+.
```
I decided to do this instead of removing the annotations because I think the annotations are useful and we should use them as well.

sipa commented on contrib/seeds/makeseeds.py:
```diff
@@ -123,34 +126,18 @@ def filtermultiport(ips: List[Dict]) -> List[Dict]:
-        sys.stderr.write(f'ERR: Could not resolve ASN for "{ip}": {e}\n')
-        return None
+def decode_ip(ip: str) -> int:
+    addr = ipaddress.ip_address(ip)
```
I believe you can instead use `asmap.net_to_prefix(ipaddress.ip_network(ip))`.

laanwj replied:
Thanks. I looked at the code briefly but wasn't under the impression `net_to_prefix` would accept individual addresses as well. Will try!

jamesob commented:
Concept ACK

sipa commented:
ACK 667e316bcb300eec131727a7ce54dd038031e267

dunxen reviewed (approved):
re-ACK 667e316

brunoerg commented:
> For now I've added instructions in README.md to download and use the demo.map from the asmap repository. When we have some other mechanism for distributing asmap files we could switch to that.

https://github.com/brunoerg/asmapy can be a good alternative!
